lib/inspec/resources/x509_certificate.rb

```diff
@@ -138,6 +145,49 @@ def extensions
       @extensions
     end
 
+    # check purpose of the certificate
+    def has_purpose?(purpose)
+      # If we have the filepath in our options we use the filepath to fetch the purposes.
+      # Else, we create a temporary file and write the content to that file.
+      # Then, use the temporary file to fetch the purposes.
+      # Todo: Check if this can be optimized or improved.
+
+      if @opts[:filepath]
+        cert_purpose = fetch_purpose(@opts[:filepath])
+      else
+        f = File.open("temporary_certificate.pem", "w")
+        f.write(@cert.to_pem)
+        f.rewind
+        begin
+          f = File.open("temporary_certificate.pem", "w")
+          f.write(@cert.to_pem)
+          f.rewind
```

Comment:
I don't understand why you are writing to the file twice with the same contents.

Reply:
Yes, it's redundant.

Reply:
@clintoncwolfe 
The command to find the certificate purpose is: `openssl x509 -noout -purpose -in [the-certificate-or-the-certificate-path]`. The `-in` argument expects a certificate file or path to a certificate file, and not content of the certificate.

There are two ways InSpec initialises a x509 certificate.
1. content
2. filepath

If the certificate was initialised with filepath, I'm using the file path directly in the command `openssl x509 -noout -purpose -in [the-certificate-or-the-certificate-path]` to fetch the certificate purpose.

If the certificate was initialised with content, I'm writing the content to a temporary file and then fetching the certificate purpose.